### PR TITLE
[air] New `train.Checkpoint` API: Update llama v2 fine-tuning example

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR updates the llama v2 example to use the new `Checkpoint` API / new persistence mode. Also does some misc cleanups.

See passing release tests here: https://buildkite.com/ray-project/release-tests-pr/builds/49582

**DON'T MERGE THIS INTO THE 2.6 TEMPLATES BRANCH.**

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py [hunk 1]
```diff
@@ -427,18 +392,72 @@ def training_function(kwargs: dict):
...
-            # only done on rank_0 which is forced to be co-located with the trainer
-            # object. By default the files created by trainer will get synced which
-            # will include the checkpoint files created by the Rank_0.
-            # Note that this will not delete the checkpoints from the previous
-            # iterations.
-            checkpoint=train.Checkpoint.from_directory(ckpt_path_epoch),
-        )
-        print("Checkpointing time: ", time.time() - checkpointing_time_s)
+        with tempfile.TemporaryDirectory() as temp_checkpoint_dir:
+            accelerator.print(f"Saving the model locally at {temp_checkpoint_dir}")
```
Comment (justinvyu) [inline doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:396]: Main change starts here

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py [hunk 2]
```diff
@@ -587,18 +607,11 @@ def main():
...
-            # can sometimes result in failed checkpoint syncing
-            # sync_config=tune.SyncConfig(sync_artifacts=False),
             storage_path=storage_path,
             checkpoint_config=train.CheckpointConfig(
                 num_to_keep=args.num_checkpoints_to_keep,
                 checkpoint_score_attribute="perplexity",
                 checkpoint_score_order="min",
-                # Enable distributed checkpointing
-                _checkpoint_keep_all_ranks=True,
-                _checkpoint_upload_from_workers=True,
```
Comment (justinvyu) [inline doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:601]: no longer need these

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py [hunk 3]
```diff
@@ -611,23 +624,20 @@ def main():
...
-    print("Results are stored in:")
-    print(results.path)
-    print("Best checkpoint is stored in:")
-    print(best_checkpoint[0].uri)
-    print(f"With perplexity: {best_checkpoint[1]['perplexity']}")
+    print("Results are stored at:")
+    print(result.path)
+    print("Best checkpoint is stored at:")
+    print(best_checkpoint)
+    print(f"With perplexity: {best_checkpoint_metrics['perplexity']}")
```
Comment (justinvyu) [inline doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:640]: I think the worst checkpoint was being reported previously 😅

Comment (kouroshHakha) [reply doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:640]: oh interesting :)

File: release/release_tests.yaml [hunk 4]
```diff
@@ -1042,7 +1042,7 @@
 
   run:
     timeout: 600
-    script: chmod +x ./run_llama_ft.sh && ./run_llama_ft.sh --size=7b --as-test
+    script: chmod +x ./run_llama_ft.sh && RAY_AIR_NEW_PERSISTENCE_MODE=1 ./run_llama_ft.sh --size=7b --as-test
```
Comment (kouroshHakha) [inline release/release_tests.yaml:1045]: do we need this flag in ./run_llama_ft.sh?

Comment (justinvyu) [reply release/release_tests.yaml:1045]: Yeah, the new codepath isn't on by default at the moment. Going to remove the need for this by 2.7, so I'm not going to update the readme. (this should NOT be merged into the workspace templates branch).

**Review by kouroshHakha — APPROVED** [review]

**Review by matthewdeng — APPROVED** [review]

**Pull request merged**